examples/fp_delegation.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <unifex/for_each.hpp>
+#include <unifex/manual_event_loop.hpp>
+#include <unifex/range_stream.hpp>
+#include <unifex/scheduler_concepts.hpp>
+#include <unifex/sync_wait.hpp>
+#include <unifex/timed_single_thread_context.hpp>
+#include <unifex/transform_stream.hpp>
+#include <unifex/via_stream.hpp>
+#include <unifex/with_query_value.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <variant>
+
+using namespace unifex;
+using namespace std::chrono_literals;
+
+namespace {
+
+class delegating_scheduler;
+
+class delegating_context {
+  public:
+  delegating_context(int capacity) : capacity_{capacity} {}
+
+  bool reserve() {
+    auto lck = std::scoped_lock{m_};
+    bool reservation_made = reservation_count_ < capacity_;
+    if(reservation_made) {
+      ++reservation_count_;
+    }
+    return reservation_made;
+  }
+
+  int get_count() {
+    return run_count_;
+  }
+
+  void run() {
+    run_count_++;
+  }
+
+  delegating_scheduler get_scheduler() noexcept;
+
+  private:
+  friend class delegating_sender;
+  std::mutex m_;
+  int reservation_count_{0};
+  std::atomic<int> run_count_{0};
+  int capacity_;
+  timed_single_thread_context single_thread_context_;
+};
+
+template <typename DelegatedOperationState, typename LocalOperationState>
+class delegating_operation final {
+  public:
+  template<class InitFunc>
+  delegating_operation(InitFunc&& func, delegating_context* context) :
+    op_{std::in_place_type_t<remove_cvref_t<decltype(func())>>{}, func()}, context_{context} {
```

Comment:
This still seems like it might require the operation returned by `func()` to support a move-constructor in order to initialise the `op_`. You'd need to, instead, pass a wrapper around `func()` that had an implicit-conversion to the return-type of `func()` so that the variant implementation would call that in a context that was direct-initialising its data member with guaranteed copy-elision.

Reply:
So... I think you're saying that we don't want to call func here, we want to pass a type that defers the call to func until the variant needs it. Is that right?

Reply:
> So... I think you're saying that we don't want to call func here, we want to pass a type that defers the call to func until the variant needs it. Is that right?

Yes. You need to pass something that is implicitly convertible to the result of calling func().
Alternatively you can use manual_lifetime_union and its .construct_from() helper to manually construct in-place.